### ISSUE TITLE
r731-733, Normal Extraction Bug

### ISSUE THREAD

**Issue by ghost** [ME3Explorer/ME3Explorer#170]
Okay, I'm reporting this now, b/c I'm pretty certain this is a bug. I mentioned it in issue #141 , and at that point I wasn't convinced, but I am now.

Texplorer and TPF Tools never had a problem previously extracting normals. For example, here's issue #56 that I reported a different problem with thumbnail normals; those extracted fine back then.

Here's what extracted normals look like now in GIMP on the right, versus when viewed in TPF Tools on the left: 

![normal_extractionbug](https://cloud.githubusercontent.com/assets/13645804/13018372/03ac50e6-d17e-11e5-9100-41fe8810983d.PNG)

To be clear, TPF Tools says these normals are perfectly fine, and they are Otte's, so I _know_ they are. The top one is unautofixed, the bottom one is autofixed. No autofixing was needed, I just thought I'd try it to see if there was a difference.

When opening both of these normals, GIMP throws a warning saying "unknown pixel format". Paint.NET won't even open them. It throws an "unrecognized file format" error.

Can anyone verify that this also happens with Photoshop? I'm testing this with Otte's EDI Overhaul From Ashes TPF, but it happens with any normal I try. It also happens just when extracting a vanilla normal from the base game via Texplorer.

**Comment by ghost** [comment]
Just wanted to add a note, that this thread on the forums might have relevant information:

http://me3explorer.freeforums.org/add-an-alpha-to-a-dxt1-texture-t2260.html

**Comment by KFreon** [comment]
This image library is pretty much based on what Microsoft says so I try to emulate how it looks.
So the following image shows Microsofts own tool opening the original (bottom) and the same one "converted" to V8U8 (as in resaved as V8U8).
![image](https://cloud.githubusercontent.com/assets/7845997/13045578/d5798406-d41f-11e5-85bf-4517616cc773.png)
Looks perfectly--wait...Something looks slightly off there.
Guess V8U8 saving is doing weird things.

Oooo...Got caught out by the basics of Computer Science. Glad there's none around...Signed vs Unsigned.
Rookie error.

**Comment by ghost** [comment]
Lol.

**Comment by ghost** [comment]
This bug was fixed, but is back. Comparison of Texplorer and GIMP:

![normbug](https://cloud.githubusercontent.com/assets/13645804/14027294/2d649f00-f1b4-11e5-83ea-b3438fb9ee42.PNG)

Also, should MIPs be extracted now, @KFreon ? I thought they should be, since you had that issue that Creeper tracked down for you. But, they are not extracted -- at the very least not in GIMP. All 3 textures I just tried have no mip layers for DXT5, DXT1, or V8U8.

**Comment by KFreon** [comment]
Mips are not extracted from the gamefiles. They could be but not without a lot of work I THINK.
It seems like the game treats each mip level as a new DDS, or at least that's how everyone elses code treats it.
Bigger images getting added get broken up into their mips, made into DDS', then installed separately.
So in theory, I could get all known mip images out, strip their DDS bits, and merge them into a coherent DDS, but I don't think that's necessary.

tldr: Mips are not extracted. 

Looking into the black thing again.

**Comment by ghost** [comment]
Mips not extracted. Got it :)

**Comment by KFreon** [comment]
OK I tested with the texture you've selected up there, and extracted right from the game it's black and white too...

**Comment by ghost** [comment]
@KFreon -- Yep, and it should not be black and white; it should look like it does in Texplorer. Normals don't convey black and white; they convey texture height, as you know.

**Comment by KFreon** [comment]
...Yes it should, but I'm saying that's a GIMP plugin problem.
That test was directly from the game files.

**Comment by ghost** [comment]
@KFreon -- Hrm. What does it look like in Photoshop? Does it look like Texplorer?

**Comment by KFreon** [comment]
Yep. It's just GIMP doing it's weird thing. 
![image](https://cloud.githubusercontent.com/assets/7845997/14063020/09e7d56a-f402-11e5-8fe7-743e5ab502ff.png)

**Comment by ghost** [comment]
@Kfreon -- Oh, fuck me.

That really, really sucks ass.

Any solution to that problem you can implement that won't totally screw everything else?

**Comment by KFreon** [comment]
I can't think of a solution at all. The code is doing exactly what it's supposed to, but GIMP is just being weird with it.

**Comment by ghost** [comment]
@Kfreon -- Could it be something like the other day with the mip count that was only problematic in GIMP and not PS?

I'll do some additional testing after dinner, but the weird thing is that I'm relatively certain this problem hasn't always existed. But, I might just be remembering different results with different normal formats, since there are several.

**Comment by KFreon** [comment]
It is possible, and I did see some weird flags there.
I'll see what I can find.

My gimp comes up with a warning about unknown pixel format sometimes though. You get that one?

**Comment by ghost** [comment]
@KFreon -- I get that sometimes, but not in this particular case.

We're trying to have dinner and watch Daredevil in the Giftfish House, so I'll be back in a bit with more info :)

**Comment by ghost** [comment]
@KFreon -- K, as I'm researching this normal map thing, can you riddle me this...

Are the game eyebrow textures, which are DXT5 _normals_: 
![brows](https://cloud.githubusercontent.com/assets/13645804/14063432/b6a7c0a8-f385-11e5-8021-d6957998901b.PNG)

Is this a result of this here:

http://tech-artists.org/wiki/Normal_map_compression#DXT5nm

**Comment by ghost** [comment]
Fuck. I just realized the problem, @KFreon.

Why I didn't check this before, I have no idea. >.<

GIMP doesn't have a V8U8 option when saving normals. If It can't save as V8U8, then it can't read V8U8.

Mystery solved.

As a fun comparison, here's a bunch of normals open in GIMP:
![normcomp](https://cloud.githubusercontent.com/assets/13645804/14063491/9a101e60-f388-11e5-90da-b02252d1c304.PNG)
1. Upper left -- DXT5 normal; the same eyebrow texture seen above in Texplorer, but open in GIMP.
2. MIddle -- ME1 and ME2 norms of the same texture. There's something a little wonky with this, though. ME1 is supposed to use ATI1N/3DC+, and ME2 is supposed to use ATI2N/3DC. There's been a recent change where all these textures for both games are labeled as ATI2N. That said, both look as they should when extracted.
3. Right, bottom -- DXT1N normal. Extracted from ME3; Joker's chair. Looks fine.
4. Right, top -- V8U8 normal of the same two textures in the middle. Obviously the format is borked, and GIMP can't read it. That really sucks. There's no plugin either, from what I can tell.

---

So, what's with the change in ME1 from ATI1N to ATI2N? Did we lose a format, like with A8R8G8B8?

**Comment by KFreon** [comment]
Ahh so GIMP is...gimped. :) :)

Also, ATI1 is a luminescence format. Black and white only, not grayscale, black and white.
Can't find any atm, but I'm pretty sure there in there somewhere.

**Comment by ghost** [comment]
@KFreon -- Damn, I swear I had this straight. I had to review some additional UDK documentation, but more importantly, I was remembering a conversation between us on the forums about a year ago...

http://me3explorer.freeforums.org/post16349.html?hilit=V8U8#p16349

Somewhere between there and here, my mind said ME1 = AT1IN, which can supposedly be used for height maps.

But, UDK clearly outlines which normal formats may be used: http://udn.epicgames.com/Three/NormalMapFormats.html. This is a References link I have on the wiki, btw.

Okay, so we technically have 4 normal formats, not 5:
1. DXT1nm (all 3 games)
2. DXT5nm (all 3 games)
3. ATI2N/3DC/BC5 (ME1/2)
4. V8U8 (ME3)

I won't make the suggestion now; I'll wait for the stable, but it would actually be nice for texture formats to be a searchable parameter. It could be handy for modders.

**Comment by ghost** [comment]
Closing this, as it's not the toolset's fault!

Stupid GIMP -_-
